include/mbgl/style/expression/collator.hpp

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <mbgl/util/feature.hpp>
+#include <mbgl/util/optional.hpp>
+
+#include <string>
+#include <memory>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+class Collator {
+public:
+    Collator(bool caseSensitive, bool diacriticSensitive, optional<std::string> locale = {});
+
+    bool operator==(const Collator& other) const;
+
+    int compare(const std::string& lhs, const std::string& rhs) const;
+
+    std::string resolvedLocale() const;
+
+    // TODO: This serialization shouldn't ever be used, but since we're part of
+    // mbgl::style::expression::Value we're expected to have a serialize()
+    mbgl::Value serialize() const;
+
+private:
+    class Impl;
+    // TODO: Figure out right copy semantics for Collator -- sharing an underlying implementation
+    // should be fine within one thread, but it might be more idiomatic to explicitly copy the
+    // implementation?
+    std::shared_ptr<Impl> impl;
```

Comment:
The one thing that comes to mind is that the Android collator has a stateful `setStrength` that theoretically could shift around during `compare` invocations (I tried an implementation in which I used two strengths to try to handle diacritic-insensitive/case-sensitive). But in the implementation I landed on, `setStrength` only happens once in the constructor.

Probably most idiomatic is using a `unique_ptr` and implementing explicit copy constructors for both `Collator` and its `Impl`s? How much copying happens to `EvaluationResult`s?

Reply:
Off the top of my head, I don't think they should be getting copied all that much, since we tend to use them like: 
 - check for error, return early if so
 - unwrap the typed value and use it for something